### ISSUE TITLE
Observed invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by edwardfung123** [argoproj/argo-workflows#7888]
## Summary

What happened/what you expected to happen? 

workflow failed but one of the template is shown as running. Upon checking the log, the controller was panic and the high level description is `runtime error: invalid memory address or nil pointer dereference`.

What executor are you using? PNS/Emissary

k8sapi

## Diagnostics

Our DAG is complex, I can't create a simplified DAG to reproduce the error yet. But it seems to me that it is semaphore related. In fact, we use the semaphore at the template-level as suggested in the doc [here](https://argoproj.github.io/argo-workflows/synchronization/#template-level-synchronization).

What Kubernetes provider are you using? `GKE 1.20.11-gke.1801`

What version of Argo Workflows are you running? `3.2.6`

The complete error and stack trace:

```
time="2022-02-15T04:57:26.663Z" level=error msg="Recovered from panic" namespace=mynamespace 
r="runtime error: invalid memory address or nil pointer dereference"
stack="
goroutine 290 [running]:
runtime/debug.Stack(0xc01e470a24, 0x1f22520, 0x331a940)
    /usr/local/go/src/runtime/debug/stack.go:24 +0x9f
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate.func2(0xc00466c480, 0x24f2c28, 0xc00012a000)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:192 +0xd1
panic(0x1f22520, 0x331a940)
    /usr/local/go/src/runtime/panic.go:971 +0x499
github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1.(*SemaphoreStatus).GetHolding(0x0, 0xc0030c7c70, 0x47, 0x47, 0x0, 0x0, 0x0, 0x0, 0x0)
    /go/src/github.com/argoproj/argo-workflows/pkg/apis/workflow/v1alpha1/workflow_types.go:2945 +0x45
github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1.(*SemaphoreStatus).LockReleased(0x0, 0xc019a11ec0, 0x55, 0xc0030c7c70, 0x47, 0x2)
    /go/src/github.com/argoproj/argo-workflows/pkg/apis/workflow/v1alpha1/workflow_types.go:2992 +0x76
github.com/argoproj/argo-workflows/v3/workflow/sync.(*Manager).Release(0xc0005b3800, 0xc004f58000, 0xc008919f80, 0x25, 0xc01d8e3d50)
    /go/src/github.com/argoproj/argo-workflows/workflow/sync/sync_manager.go:195 +0x343
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc00c7a25a0, 0xeb, 0x24dbf80, 0xc005642960, 0xc03040c9c0, 0xc005633540, 0x2, ...)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1656 +0x534
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc0145f99d0, 0xc0297e7740, 0xa)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:483 +0x90b
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc0145f99d0, 0xc0297e7770, 0xf)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:408 +0x4a5
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc0145f99d0, 0xc0297e7737, 0x9)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:408 +0x4a5
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAG(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc03fdb0600, 0x3d, 0xc03040c9c0, 0xc030ce3da8, 0x6, 0xc033dccb40, 0x24dbf80, ...)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:244 +0x309
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc03fdb0600, 0x3d, 0x24dbf80, 0xc005624690, 0xc03168bbc0, 0xc0055f9900, 0x1, ...)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1806 +0x16bc
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc0145f97a0, 0xc030ce3cc0, 0xd)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:483 +0x90b
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAG(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc010b9c3f0, 0x2f, 0xc03168bbc0, 0xc030ce3a10, 0x6, 0xc024648b40, 0x24dbf80, ...)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:244 +0x309
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc010b9c3f0, 0x2f, 0x24dbf80, 0xc0055efc20, 0xc03168b9c0, 0xc0055f8c80, 0x3, ...)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1806 +0x16bc
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc0145f9570, 0xc019a10f0a, 0x5)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:483 +0x90b
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc0145f9570, 0xc030ce38a0, 0x6)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:408 +0x4a5
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAG(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc010b8d1d0, 0x29, 0xc03168b9c0, 0xc030ce34c4, 0x6, 0xc02f22d8c0, 0x24dbf80, ...)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:244 +0x309
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc010b8d1d0, 0x29, 0x24dbf80, 0xc0055ee690, 0xc03168b6c0, 0xc0055f4000, 0x5, ...)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1806 +0x16bc
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc0145f92d0, 0xc030ce3390, 0xe)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:483 +0x90b
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc0145f92d0, 0xc0072cd6e5, 0x13)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:408 +0x4a5
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc0145f92d0, 0xc030ce33a0, 0x10)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:408 +0x4a5
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAG(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc031c213a0, 0x1a, 0xc03168b6c0, 0xc030ce3098, 0x6, 0xc0176998c0, 0x24dd0f0, ...)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:244 +0x309
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate(0xc00466c480, 0x24f2c28, 0xc00012a000, 0xc031c213a0, 0x1a, 0x24dd0f0, 0xc00466c540, 0xc03168b680, 0xc004f2ee60, 0x2, ...)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1806 +0x16bc
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate(0xc00466c480, 0x24f2c28, 0xc00012a000)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:344 +0xf3e
github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).processNextItem(0xc000477400, 0x24f2c28, 0xc00012a000, 0x0)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:770 +0x945
github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).runWorker(0xc000477400)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:692 +0x9b
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc006c00e40)
    /go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc006c00e40, 0x24ab6c0, 0xc006966a80, 0x1, 0xc00e3a6600)
    /go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:156 +0x9b
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc006c00e40, 0x3b9aca00, 0x0, 0x1, 0xc00e3a6600)
    /go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc006c00e40, 0x3b9aca00, 0xc00e3a6600)
    /go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:90 +0x4d
created by github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).startLeading
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:309 +0x48a
" 
workflow=myworkflow
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by edwardfung123** [comment]
@sarabala1979 Sorry I upgraded the server to `3.2.7` to see if it fixes the issue. I still get the nil pointer error. But the stack trace showed a different error. It changed from `SemaphoreStatus` to `MutexStatus`.

Here is the workflow for `3.2.7`:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: synchronization-tmpl-level-
spec:
  entrypoint: dag-1
  templates:
  - name: dag-1
    dag:
      failFast: false
      tasks:
      - name: dag-1-task-1
        template: django-command
      - depends: dag-1-task-1
        name: dag-1-task-2
        template: dag-2
      - depends: dag-1-task-2
        name: dag-1-task-3
        template: dag-3

  - name: load-command
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["S=$(shuf -i 2-2 -n 1); echo begin to sleep $S; sleep $S; echo acquired lock"]
      resources:
        requests:
          cpu: "0.1"
          memory: 100Mi

  - name: django-command
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["echo 'django command!'"]
      resources:
        requests:
          cpu: "0.1"
          memory: 100Mi

  - name: dag-2
    dag:
      tasks:
      - name: dag-2-task-1  
        template: django-command
    synchronization:
      mutex:
        name: dag-2-task-1

  - name: dag-3
    dag:
      failFast: false
      tasks:
      - name: dag-3-task-1
        template: django-command
      - depends: dag-3-task-1
        name: dag-3-task-2
        template: django-command
      - depends: dag-3-task-2
        name: dag-3-task-3
        template: django-command
      - depends: dag-3-task-3
        name: dag-3-task-4
        template: dag-4
      - depends: dag-3-task-4.Succeeded
        name: dag-3-task-5
        template: django-command
      - depends: >
          dag-3-task-3
          && (dag-3-task-4.Skipped || dag-3-task-4.Failed || dag-3-task-4.Errored 
          || dag-3-task-5.Failed
          || dag-3-task-5.Errored)
        name: failed
        template: django-command

  - name: dag-4
    dag:
      failFast: false
      tasks:
      - name: dag-4-task-1
        template: django-command
      - depends: dag-4-task-1
        name: dag-4-task-2
        template: dag-5

  - name: dag-5
    dag:
      failFast: false
      tasks:
      - name: dag-5-task-1
        template: django-command
      - depends: dag-5-task-1
        name: dag-5-task-2
        template: django-command
      - depends: dag-5-task-2
        name: dag-5-task-3
        template: django-command
      - depends: dag-5-task-3
        name: dag-5-task-4
        template: django-command
      - depends: dag-5-task-4
        name: dag-5-task-5
        template: dag-6
        withParam: '["1","2"]'
      - depends: dag-5-task-5
        name: dag-5-task-6
        template: django-command
      - depends: dag-5-task-6
        name: dag-5-task-7
        template: django-command
    
  - name: dag-6
    dag:
      tasks:
      - name: dag-6-task-1
        template: load-command
    synchronization:
      semaphore:
        configMapKeyRef:
          name: test-config-map
          key: test-sem
```

The stacktrace:

```
r="runtime error: invalid memory address or nil pointer dereference"
stack="
goroutine 277 [running]:
runtime/debug.Stack(0xc065ea75bb, 0x1f22520, 0x331a940)
    /usr/local/go/src/runtime/debug/stack.go:24 +0x9f
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate.func2(0xc002df9a40, 0x24f2de8, 0xc000058018)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:192 +0xd1
panic(0x1f22520, 0x331a940)
    /usr/local/go/src/runtime/panic.go:971 +0x499
github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1.(*MutexStatus).GetHolding(0x0, 0xc00bdbed50, 0x26, 0xc006fe9f60, 0x1e3ed00, 0xc006fe9f70, 0xc000079000, 0x0)
    /go/src/github.com/argoproj/argo-workflows/pkg/apis/workflow/v1alpha1/workflow_types.go:3039 +0x26
github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1.(*MutexStatus).LockReleased(0x0, 0xc00d4ce380, 0x62, 0xc00bdbed50, 0x26, 0x2)
    /go/src/github.com/argoproj/argo-workflows/pkg/apis/workflow/v1alpha1/workflow_types.go:3092 +0x4c
github.com/argoproj/argo-workflows/v3/workflow/sync.(*Manager).Release(0xc0000fcae0, 0xc00626f800, 0xc0078625d0, 0x2b, 0xc006fe9e60)
    /go/src/github.com/argoproj/argo-workflows/workflow/sync/sync_manager.go:195 +0x343
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc002df9a40, 0x24f2de8, 0xc000058018, 0xc0073183f0, 0xc007835b10, 0x10)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:358 +0x1fe5
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0xc002df9a40, 0x24f2de8, 0xc000058018, 0xc0073183f0, 0xc006f810c8, 0x12)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:408 +0x4a5
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAG(0xc002df9a40, 0x24f2de8, 0xc000058018, 0xc0107569e0, 0x20, 0xc007d47240, 0xc00bdbeb40, 0x26, 0xc03286dd40, 0x24dd2b0, ...)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:244 +0x309
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate(0xc002df9a40, 0x24f2de8, 0xc000058018, 0xc0107569e0, 0x20, 0x24dd2b0, 0xc002df9b00, 0xc007d47200, 0x0, 0x0, ...)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:1807 +0x16bc
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate(0xc002df9a40, 0x24f2de8, 0xc000058018)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:344 +0xf3e
github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).processNextItem(0xc000101800, 0x24f2de8, 0xc000058018, 0x0)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:770 +0x945
github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).runWorker(0xc000101800)
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:692 +0x9b
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc00d3aeeb0)
    /go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00d3aeeb0, 0x24ab840, 0xc00d00f7a0, 0x1, 0xc00c6831a0)
    /go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:156 +0x9b
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00d3aeeb0, 0x3b9aca00, 0x0, 0x301, 0xc00c6831a0)
    /go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc00d3aeeb0, 0x3b9aca00, 0xc00c6831a0)
    /go/pkg/mod/k8s.io/apimachinery@v0.21.5/pkg/util/wait/wait.go:90 +0x4d
created by github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).startLeading
    /go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:309 +0x48a
"
```

**Comment by edwardfung123** [comment]
I re-ran the reproducible workflow in `3.2.6`. I got the same `MutexStatus` error.  I hope it gives you enough info to identify and fix the bug.

**Comment by louisnow** [comment]
Similar error when using an exit handler
https://github.com/argoproj/argo-workflows/issues/7934

**Comment by sarabala1979** [comment]
@edwardfung123 Issue fixed and released in 3.3.0-rc8. Can you test it?
